designs/v1alpha6.md

```diff
@@ -0,0 +1,99 @@
+# v1alpha6 API Proposal
+
+This document formalizes the [v1alpha6 laundry list](https://github.com/aws/karpenter/issues/1327) into a release strategy and concrete set of proposed changes.
+
+### Migration path from v1alpha5 to v1alpha6
+
+Customers will be able to migrate from v1alpha5 to v1alpha6 in a single cluster using a single Karpenter version.
+
+Kubernetes custom resources have built in support for API version compatibility. CRDs with multiple versions must define a “storage version”, which controls the data stored in etcd. Other versions are views onto this data and converted using [conversion webhooks](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion). However, there is a fundamental limitation that all versions must be safely [round-tripable through each other](https://book.kubebuilder.io/multiversion-tutorial/api-changes.html). This means that it must be possible to define a function that converts a v1alpha5 Provisioner into a v1alpha6 Provisioner and vise versa.
+
+Unfortunately, multiple proposed changes in v1alpha6 are not round-trippable. Below, we propose two deprecations of legacy fields in favor more modern mechanisms that have seen broad adoption in v1alpha5. These changes remove sharp edges that regularly cause customer surprise and production pain.
+
+To work around this limitation, we have three options:
+
+1. Rename Provisioner to something like NodeProvisioner, to avoid being subject to round-trippable requirements
+2. [Recommended] Require that users delete the existing v1alpha5 Provisioner CRD and then install the v1alpha6 Provisioner CRD. This will result in all capacity being shut down, and cannot be done in place if the customer has already launched nodes.
+3. Keep the legacy fields in our API forever
+
+Option 2 is untenable and easily discarded. We must provide a migration path for existing customers. Option 3 minimizes immediate impact to customers, but results in long term customer pain. There are other benefits to renaming Provisioner, so while it does cause some churn, it results in long term value.
+
+Following option #1, customers would upgrade as follows:
+
+1. Install Karpenter version 0.x.0, this version supports both v1alpha5.Provisioner and v1alpha6.NodeProvisioner
+    1. We may choose to support both APIs in more than one Karpenter version at the cost of developer velocity
+2. Create a v1alpha6.NodeProvisioner, and add a taint to the old v1alpha5.Provisioner
```

Comment:
Could the new Karpenter version automate that transition? I think it's technically possible.

If not, we could provide a container image that runs a one-time migration job (either in-cluster with appropriate access, or to run locally).